### PR TITLE
Fix DBAL MySQL tests (ORM + Jackalope). Introduce Sqlite + Postgres to matrix.

### PR DESCRIPTION
**Pull request by dantleech**
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | https://github.com/sulu/sulu/issues/1257
| License | MIT

#### What's in this PR?

This PR will fix the Doctrine DBAL PHPCR tests (for both the ORM and Jackalope). It also (re)introduces sqlite and postgres to the test matrix (as allowed fails).

#### Why?

Because we should test Sulu with the more commonly used DBAL Jackalope implementation and not only
the Jackalope Jackrabbit PHPCR implementation.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -13,18 +9,52 @@ cache:
...
+    - php: 7.0
+      env: 
+        - SYMFONY__PHPCR__TRANSPORT=jackrabbit
+
+        # restart jackrabbit after each suite see: https://github.com/sulu-io/sulu/issues/2137
+        - TEST_FLAGS="--jackrabbit-restart"
+    - php: 7.0
+      env:
+        - SYMFONY__DATABASE__DRIVER=pdo_sqlite
+        - SYMFONY__DATABASE__PATH=/tmp/test.sqlite
```
Comment (dantleech) [inline .travis.yml]: this can be changed to `test.sqlite` i think.

File: .travis.yml [hunk 2]
```diff
@@ -13,18 +9,52 @@ cache:
...
 
 before_script:
-  - wget http://archive.apache.org/dist/jackrabbit/$JACKRABBIT_VERSION/jackrabbit-standalone-$JACKRABBIT_VERSION.jar
-  - java -jar jackrabbit-standalone-$JACKRABBIT_VERSION.jar > /dev/null &
+  - |
+    if [[ $SYMFONY__PHPCR__TRANSPORT = jackrabbit ]]; then 
+      wget http://archive.apache.org/dist/jackrabbit/$JACKRABBIT_VERSION/jackrabbit-standalone-$JACKRABBIT_VERSION.jar
+      java -jar jackrabbit-standalone-$JACKRABBIT_VERSION.jar > /dev/null &
+    fi
+  - echo "memory_limit=2048M" >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
```
Comment (dantleech) [inline .travis.yml]: add comment aboout why this is required.

File: bin/runtests [hunk 3]
```diff
@@ -324,7 +334,7 @@ function init_dbal()
     write_info('Creating database');
     exec_sf_cmd(
         'doctrine:database:create',
-        false, false
+        true, true
```
Comment (dantleech) [inline bin/runtests]: Revert?

File: src/Sulu/Bundle/ContentBundle/Tests/Functional/Controller/NodeControllerTest.php [hunk 4]
```diff
@@ -1242,10 +1242,18 @@ public function testSmartContent()
...
+        if ($this->isDatabaseDriver('PDOSqlite')) {
+            $this->markTestSkipped('Jackalope Doctrine DBAL with the Sqlite driver does not properly compare multi-fields');
+            return;
+        }
+
         $client = $this->createAuthenticatedClient();
 
         $client->request('GET', '/api/nodes/filter?webspace=sulu_io&language=en');
         $response = json_decode($client->getResponse()->getContent(), true);
+        $this->assertHttpStatusCode(200, $client->getResponse());
```
Comment (dantleech) [inline src/Sulu/Bundle/ContentBundle/Tests/Functional/Controller/NodeControllerTest.php]: This test should really be refactored to the provider pattern. however, its OOS

File: src/Sulu/Bundle/CustomUrlBundle/Resources/config/document.xml [hunk 5]
```diff
@@ -26,7 +26,7 @@
             <argument type="service" id="sulu_core.webspace.webspace_manager"/>
             <argument type="service" id="sulu.phpcr.session"/>
 
-            <tag name="sulu_document_manager.initializer"/>
+            <tag name="sulu_document_manager.initializer" priority="-127"/>
```
Comment (dantleech) [inline src/Sulu/Bundle/CustomUrlBundle/Resources/config/document.xml]: Add comment.

File: src/Sulu/Bundle/MediaBundle/Collection/Manager/CollectionManager.php [hunk 6]
```diff
@@ -527,6 +527,11 @@ public function delete($id)
             throw new CollectionNotFoundException($id);
         }
 
+        // SQlite does not support cascade-delete, so manually delete the medias.
+        foreach ($collectionEntity->getMedia() as $media) {
+            $this->em->remove($media);
+        }
```
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Collection/Manager/CollectionManager.php]: Remove this, we are not going to emulate the RDBMS - although maybe not required as I think the ORM should be able to "soft cascade": `cascade: [ "delete" ]`, perhaps.

File: src/Sulu/Bundle/TestBundle/Testing/SuluTestCase.php [hunk 7]
```diff
@@ -207,4 +219,30 @@ protected function getEntityManager()
...
+        if (!class_exists($class)) {
+            throw new \Exception(sprintf(
+                'Checking to see if %s is being used, but driver class "%s" does not exist',
+                $driverName, 
+                $class
+            ));
+        }
+
+        return $this->getContainer()->get('doctrine')->getConnection()->getDriver() instanceof $class;
+    }
```
Comment (dantleech) [inline src/Sulu/Bundle/TestBundle/Testing/SuluTestCase.php]: Remove this if we decide to allow sqlite failures/

File: src/Sulu/Component/Content/Tests/Functional/Repository/ContentRepositoryTest.php [hunk 8]
```diff
@@ -637,11 +637,32 @@ public function testFindWithNonFallbackProperties()
...
+        $this->assertGreaterThan($link->getCreated(), $created);
+        $this->assertGreaterThan($link->getChanged(), $changed);
+
+        // Jackalope Doctrine DBAL will currently "normalize" to UTC when
+        // persisting, but load in the default TZ on loading - normalizing to
+        // UTC means the dates are equivilent but not the same for the purposes
+        // of comparison.
+        //
+        // Pending PR: https://github.com/jackalope/jackalope-doctrine-dbal/pull/326
+        $page = $this->documentManager->find($page->getUuid(), 'de');
```
Comment (dantleech) [inline src/Sulu/Component/Content/Tests/Functional/Repository/ContentRepositoryTest.php]: This has been merged, need to wait until no release before removing this hack.

Comment (dantleech) [inline src/Sulu/Component/Content/Tests/Functional/Repository/ContentRepositoryTest.php]: Also, explain why we reload (because reloading will cause the date to be updated to the correct value).

File: src/Sulu/Component/Content/SmartContent/QueryBuilder.php [hunk 9]
```diff
@@ -286,7 +286,7 @@ private function buildTagsWhere($tags, $operator, $languageCode)
             );
 
             foreach ($tags as $tag) {
-                $sql2Where[] = 'page.[' . $property->getName() . '] = ' . $tag;
+                $sql2Where[] = 'page.[' . $property->getName() . '] = ' . $tag . '';
```
Comment (dantleech) [inline src/Sulu/Component/Content/SmartContent/QueryBuilder.php]: Remove `''`

File: .travis.yml [hunk 10]
```diff
@@ -13,18 +9,52 @@ cache:
...
   global:
     - JACKRABBIT_VERSION=2.12.0
+    - SYMFONY__PHPCR__TRANSPORT=doctrinedbal
+    - TEST_FLAGS=""
+
+matrix:
+  include:
+    - php: 5.5
+      env: 
+        - SYMFONY__DATABASE__DRIVER=pdo_mysql
```
Comment (dantleech) [inline .travis.yml]: this is the default, can remove it here.

File: .travis.yml [hunk 11]
```diff
@@ -13,18 +9,53 @@ cache:
...
   global:
     - JACKRABBIT_VERSION=2.12.0
+    - SYMFONY__PHPCR__TRANSPORT=doctrinedbal
+    - TEST_FLAGS=""
+    - SYMFONY__DATABASE__DRIVER=pdo_mysql
+
+matrix:
+  include:
+    - php: 5.5
+      env:
```
Comment (dantleech) [inline .travis.yml]: dd

File: .travis.yml [hunk 12]
```diff
@@ -13,18 +9,53 @@ cache:
...
+    - SYMFONY__DATABASE__DRIVER=pdo_mysql
+
+matrix:
+  include:
+    - php: 5.5
+      env: 
+    - php: 7.0
+      env: 
+        - SYMFONY__PHPCR__TRANSPORT=jackrabbit
+
```
Comment (dantleech) [inline .travis.yml]: dd

File: bin/runtests [hunk 13]
```diff
@@ -399,11 +411,19 @@ function init_bundle($phpunitPath)
     }
 
     write_info('Doctrine ORM detected, updating the schema for the bundle.');
+
+    if ($input->getOption('database-update-drop')) {
+        exec_cmd(sprintf('%s %s', $console, 'doctrine:database:drop --force'));
+        exec_cmd(sprintf('%s %s', $console, 'doctrine:database:create'));
+    }
```
Comment (dantleech) [inline bin/runtests]: This is not required now since we do not use foreign keys with Sqlite. Remove what is not required ?

File: bin/runtests [hunk 14]
```diff
@@ -449,7 +469,7 @@ if ($input->getOption('initialize')) {
     write_header('Initialization');
     init_dbal();
     write_info('Initializing the content repository');
-    exec_sf_cmd('sulu:document:initialize --ansi', false);
+    exec_sf_cmd('sulu:document:initialize --ansi', true);
     if (getenv('SYMFONY__PHPCR__TRANSPORT') === 'doctrine_dbal') {
         init_phpcr_dbal();
```
Comment (dantleech) [inline bin/runtests]: (general) Force colors on PHPUnit.

File: src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php [hunk 15]
```diff
@@ -214,20 +214,31 @@ public function testCGet()
...
+
+        foreach ($categories as $category) {
+            if (isset($category['name']) && $category['name'] === 'First Category') {
+                $firstCategory = $category;
+            }
+
+            if ($category['key'] === 'second-category-key') {
+                $secondCategory = $category;
+            }
+        }
```
Comment (dantleech) [inline src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php]: Add comment indicating that the results are unordered.

File: src/Sulu/Bundle/MediaBundle/Tests/Unit/Entity/MediaTest.php [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+    /**
+     * It should throw an exception if the Media is not associated with a collection
+     * and getCollection is called.
+     *
+     * @expectedException \RuntimeException
+     */
+    public function testGetCollectionNull()
+    {
+        $this->media->getCollection();
```
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Tests/Unit/Entity/MediaTest.php]: I assume that a media not having a collection is an invalid state. /cc @alexander-schranz

File: src/Sulu/Component/Content/Tests/Functional/Repository/ContentRepositoryTest.php [hunk 17]
```diff
@@ -637,11 +637,39 @@ public function testFindWithNonFallbackProperties()
...
+
+        $this->assertGreaterThan($link->getCreated(), $created);
+        $this->assertGreaterThan($link->getChanged(), $changed);
+
+        // Reload the document.
+        //
+        // Jackalope Doctrine DBAL will currently "normalize" to UTC when
+        // persisting, but load in the default TZ on loading - normalizing to
+        // UTC means the dates are equivilent but not the same for the purposes
+        // of comparison.
```
Comment (dantleech) [inline src/Sulu/Component/Content/Tests/Functional/Repository/ContentRepositoryTest.php]: This comment is not acurate.

**Comment by dantleech** [comment]
Tests are now passing. I have decided to let the Sqlite tests fail and ignore them.

Sqlite doesn't work properly, and we shouldn't hide that. I will create an issue detailing the problems that need to be addressed.

File: .travis.yml [hunk 18]
```diff
@@ -13,18 +9,51 @@ cache:
...
+    - php: 7.0
+      env:
+        - SYMFONY__DATABASE__DRIVER=pdo_sqlite
+        - SYMFONY__DATABASE__PATH=test.sqlite
+    - php: 7.0
+      env: 
+        - SYMFONY__DATABASE__DRIVER=pdo_pgsql
+        - SYMFONY__DATABASE__USER=postgres
+        - SYMFONY__DATABASE__PASSWORD=postgres
+  allow_failures:
```
Comment (danrot) [inline .travis.yml]: I would not add these... Nobody looks at them anyway, they won't work if we do something about it, and it delays the build process a lot.

File: .travis.yml [hunk 19]
```diff
@@ -13,18 +9,51 @@ cache:
...
+    - php: 5.5
+    - php: 7.0
+      env: 
+        - SYMFONY__PHPCR__TRANSPORT=jackrabbit
+        # restart jackrabbit after each suite see: https://github.com/sulu-io/sulu/issues/2137
+        - TEST_FLAGS="--jackrabbit-restart"
+    - php: 7.0
+      env:
+        - SYMFONY__DATABASE__DRIVER=pdo_sqlite
+        - SYMFONY__DATABASE__PATH=test.sqlite
```
Comment (danrot) [inline .travis.yml]: SQLite is working now?

File: src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php [hunk 20]
```diff
@@ -214,20 +214,32 @@ public function testCGet()
 
         $this->assertHttpStatusCode(200, $client->getResponse());
 
-        $response = json_decode($client->getResponse()->getContent());
+        $response = json_decode($client->getResponse()->getContent(), true);
+
+        $categories = $response['_embedded']['categories'];
+
+        // results are unordered, and order is inconsistent between database drivers
+        foreach ($categories as $category) {
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php]: Why not using something like `usort`? The logic you've added here gets quite complicated once a few more categories are added.

File: .travis.yml (hunk 19 above)
Comment (dantleech) [inline .travis.yml]: No, its an allowed failure, I have listed the issues we are facing here:
https://github.com/sulu/sulu/issues/2048

File: src/Sulu/Bundle/DocumentManagerBundle/Initializer/WorkspaceInitializer.php [hunk 21]
```diff
@@ -39,13 +39,9 @@ public function initialize(OutputInterface $output)
...
             try {
-                $connection->getRootNode();
+                $workspace->createWorkspace($workspace->getName());
                 $output->writeln(sprintf('  [ ] <info>workspace</info>: "%s"', $workspace->getName()));
             } catch (RepositoryException $e) {
-                // TODO: We should catch the more explicit
-                // WorkspaceNotFoundException but Jackalope doctrine-dbal does
-                // not throw this: https://github.com/jackalope/jackalope-doctrine-dbal/issues/322
-                $workspace->createWorkspace($workspace->getName());
                 $output->writeln(sprintf('  [+] <info>workspace</info>: "%s"', $workspace->getName()));
```
Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/Initializer/WorkspaceInitializer.php]: Should this line really be printed in case of error? If yes I would move it after the try/catch block, because the same line is in the try and catch.

File: src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php (hunk 20 above)
Comment (dantleech) [inline src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php]: Yes, looking at the postgres tests, that would indeed be better.

File: src/Sulu/Bundle/MediaBundle/Entity/Media.php [hunk 22]
```diff
@@ -172,6 +172,13 @@ public function setCollection(\Sulu\Bundle\MediaBundle\Entity\CollectionInterfac
      */
     public function getCollection()
     {
+        if (null === $this->collection) {
+            throw new \RuntimeException(sprintf(
+                'Media entity %d is not associated with a collection',
+                $this->getId()
+            ));
+        }
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Entity/Media.php]: Why is that needed? That should actually never be the case.

File: src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php [hunk 23]
```diff
@@ -851,21 +850,26 @@ public function delete($id, $checkSecurity = false)
...
 
-            $mediaEntity->setCollection($this->em->getReference(self::ENTITY_NAME_COLLECTION, $destCollection));
+        if ($mediaEntity === null) {
+            throw new MediaNotFoundException($id);
+        }
 
-            $this->em->flush();
+        // use find rather than ->getReference as its behavior when a collection doesn't
+        // exist is universal (returns NULL) and using a proxy provides no worth-while
+        // performance benefit for move.
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php]: Are you sure? From what I understand it won't even hydrate the object, so it should have a better performance fore sure.

File: src/Sulu/Bundle/DocumentManagerBundle/Initializer/WorkspaceInitializer.php (hunk 21 above)
Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Initializer/WorkspaceInitializer.php]: Some duplcation here .. basically if there was an RepositoryException then the workspace does NOT exist, so it will be `[+]` added, if no exception was thrown then it already exists (`[ ]` - no change).

Comment (dantleech) [inline src/Sulu/Bundle/DocumentManagerBundle/Initializer/WorkspaceInitializer.php]: Strike that, it should be reversed.

File: src/Sulu/Bundle/MediaBundle/Entity/Media.php (hunk 22 above)
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Entity/Media.php]: Because it is NOT the case if you use Sqlite and delete the collection. The media exists and it has no collection. In anycase anything that should ALWAYS return an object should throw an exception if it is an invalid state (otherwise you get a Fatal Error).

I am assuming that no media entity should exist without a collection.

File: src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php (hunk 23 above)
Comment (dantleech) [inline src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php]: Hmm, I guess the issue is actually that, when `flush()` was being called, the foreign key constraint was failing and throwing the DBALException. As Sqlite had no foreign key support, this was failing.

I will revert this as I think we are working to a policy of not working-around Sqlite foreign key support.

File: src/Sulu/Bundle/TranslateBundle/Controller/CodeController.php [hunk 24]
```diff
@@ -279,8 +279,24 @@ public function putAction(Request $request, $id)
...
+            }
+
+            if (null === $location) {
+                throw new \InvalidArgumentException(sprintf(
+                    'Location "%s" not found', $location['id']
+                ));
+            }
+
+            $code->setPackage($package);
+            $code->setLocation($location);
```
Comment (dantleech) [inline src/Sulu/Bundle/TranslateBundle/Controller/CodeController.php]: Not sure if this should be kept. Although this was a Sqlite work-around, I do not think it is a *bad* thing to explicitly check for the existence of the package / location. The error response is clearer and this is not performance critical.

File: .travis.yml [hunk 25]
```diff
@@ -13,20 +9,48 @@ cache:
...
+#    - php: 7.0
+#      env:
+#        - SYMFONY__DATABASE__DRIVER=pdo_sqlite
+#        - SYMFONY__DATABASE__PATH=test.sqlite
+#    - php: 7.0
+#      env: 
+#        - SYMFONY__DATABASE__DRIVER=pdo_pgsql
+#        - SYMFONY__DATABASE__USER=postgres
+#        - SYMFONY__DATABASE__PASSWORD=postgres
+#  fast_finish: true # do not wait for allowed failures, as they do not affect the build.
```
Comment (dantleech) [inline .travis.yml]: Commented rather than deleted, as I think some of this is non-evident and it is useful for testing in the future.

**Comment by danrot** [comment]
@dantleech Damn, forgot about this one... I guess you have to rebase it because of the other testing pr.

File: bin/runtests [hunk 26]
```diff
@@ -182,14 +189,15 @@ function show_help()
...
-  -C, --no-component-tests  Do not run the component tests.
-  -B, --no-bundle-tests     Do not run the bundle tests.
+  -i, --initialize            Execute the initializaction script before running the tests.
+  -f, --flags                 Pass flags to PHPUnit.
+  -t, --target                Specify a target bundle.
+  -a, --all                   Run all tests.
+  -r, --restart-jackrabbit    Restart jackrabbit between bundle tests.
+  -h, --help                  Show this help.
+  -C, --no-component-tests    Do not run the component tests.
+  -B, --no-bundle-tests       Do not run the bundle tests.
```
Comment (danrot) [inline bin/runtests]: Shouldn't we turn that logic? I think if the command is executed with the `-t` flag I never want the bundle tests executed.

**Comment by danrot** [comment]
@dantleech Can you fix the conflicts here now? Then we could merge that, since the release is done now.

File: bin/runtests (hunk 26 above)
Comment (dantleech) [inline bin/runtests]: Maybe, we could probably refactor this to be much better, but out of scope here I think.

**Comment by dantleech** [comment]
Updated, will need to wait for the tests as made some changes to the .travis file.
